lib/aws-record/record/batch_read.rb

```diff
@@ -0,0 +1,154 @@
+
+module Aws
+  module Record
+    class BatchRead
+
+      BATCH_GET_ITEM_LIMIT = 100
+
+      def initialize(opts = {})
+        @client = opts[:client]
+      end
+
+      def find(klass, key = {})
+        unprocessed_key = format_unprocessed_key(klass, key)
+        store_unprocessed_key(klass, unprocessed_key)
+        store_item_class(klass, unprocessed_key)
+      end
+
+      def execute!
+        if unprocessed_keys.count > BATCH_GET_ITEM_LIMIT
+          operation_keys = unprocessed_keys[0..BATCH_GET_ITEM_LIMIT-1].dup
+          @unprocessed_keys = unprocessed_keys[BATCH_GET_ITEM_LIMIT..-1].dup
+        else
+          operation_keys = unprocessed_keys.dup
+          @unprocessed_keys.clear
+        end
+
+        operations = build_operations(operation_keys)
+        result = @client.batch_get_item(request_items: operations)
+        build_items(result.responses)
+
+        unless result.unprocessed_keys.nil?
+          update_unprocessed_keys(result.unprocessed_keys)
+        end
+
+        self
+      end
+
+      def complete?
+        unprocessed_keys.none?
+      end
+
+      def items
+        @items ||= []
+      end
+
+      private
+      def unprocessed_keys
+        @unprocessed_keys ||=[]
+      end
+
+      def item_classes
+        @item_classes ||= {}
+      end
+
+      def format_unprocessed_key(klass, key)
+        item_key = {}
+        attributes = klass.attributes
+        klass.keys.each_value do |attr_sym|
+          unless key[attr_sym]
+            raise Errors::KeyMissing.new(
+              "Missing required key #{attr_sym} in #{key}"
+            )
+          end
+          attr_name = attributes.storage_name_for(attr_sym)
+          item_key[attr_name] = attributes.attribute_for(attr_sym).
+            serialize(key[attr_sym])
+        end
+        item_key
+      end
+
+      def store_unprocessed_key(klass, unprocessed_key)
+        unprocessed_keys << {:keys => unprocessed_key, :table_name => klass.table_name}
+      end
+
+      def store_item_class(klass, key)
+        if item_classes.include?(klass.table_name)
+          item_classes[klass.table_name].each do |item|
+            if item[:keys] == key && item[:class] != klass
+              raise 'Provided item keys is a duplicate request'
+            end
+          end
+        end
+        item_classes[klass.table_name] ||= []
+        item_classes[klass.table_name] << {keys: key, class: klass}
+      end
+
+      def build_operations(keys)
+        operations = {}
+        keys.each do | key |
+          operations[key[:table_name]] ||= { keys: [] }
+          operations[key[:table_name]][:keys] << key[:keys]
+        end
+        operations
+      end
+
+      def build_items(item_responses)
+        item_responses.each do |table, unprocessed_items|
+          unprocessed_items.each do |item|
+            item_class = find_item_class(table, item)
+            if item_class.nil?
+              raise 'Item Class was not found'
```

Comment:
This is a tricky case.  This would only happen if we get an unexpected response from the service.  I'm trying to think through what we should do here / what a user would want to do with this case.
One option is to log a warning but keep going - if we got an extra item in response, but all of the other items we expected - then we could still keep constructing all of those.  

On the other hand, I really don't know how this would happen and it could indicate something was maybe truly wrong.  

I think overall I'd lean towards logging it as a warning but moving on rather than raising an error.

Reply:
Completed! Per our conversation on Friday, we will be logging a warning and will process all the remaining items.

Reply:
Also - added unit test to cover this.